### PR TITLE
fix: Add source map validation for webpack

### PR DESCRIPTION
**Pull request by david0xd**
## **Description**
This PR adds source map validation suitable for webpack builds.

Original idea is taken from the [previous implementation](https://github.com/MetaMask/metamask-extension/blob/main/development/sourcemap-validator.js#L1) used for browserify builds.

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/40025?quickstart=1)

## **Changelog**

CHANGELOG entry: null

## **Related issues**
Fixes: https://github.com/MetaMask/MetaMask-planning/issues/6363

## **Manual testing steps**
1. Build MetaMask with `yarn webpack` or `yarn webpack --targetEnvironment production`.
2. Execute following command in terminal: `yarn validate-source-maps:webpack`.
3. Observe result. All source maps should be valid and there should be no errors in console.
4. Go to some of the source files and make some changes (add few new lines).
5. Run command for validation again: `yarn validate-source-maps:webpack`.
6. Observe result. There should be errors printed out in the terminal telling more about wrong source map.

### **Before**
Source map validation for webpack was not available before. Nothing to show here.

#### Valid source maps example
Screenshot 2026-02-18 at 15 07 54
Screenshot 2026-02-18 at 15 08 29

#### Invalid source maps example
Screenshot 2026-02-18 at 15 46 17

## **Pre-merge author checklist**
- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**
- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

---

> [!NOTE]
> **Medium Risk**
> Adds a new CI step that can fail builds based on webpack sourcemap correctness; risk is mainly around potential false positives/negatives affecting build reliability rather than runtime behavior.
> 
> **Overview**
> Adds a dedicated webpack sourcemap validator (`yarn validate-source-maps:webpack`) that scans `dist/chrome` for `.js`+`.map` pairs and verifies mappings for `new Error` callsites (with comment-block skipping and clearer code-frame error output).
> 
> Wires this into CI by extending `run-build.yml` with an optional `validate-source-maps-webpack` input/step, and includes a comprehensive `node:test` suite plus fixtures for validator edge cases; updates deps/lockfile for `@types/babel__code-frame`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 7c31f1c494186cee0f3df41bc8d7083257c99a64. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: development/webpack/sourcemap-validator.ts [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+      const sourceLine = sourceLines[sourceLineIndex];
+      const column = result.column ?? 0;
+      const portion = sourceLine ? sourceLine.slice(column) : '';
+      const foundValidSource = portion.includes(TARGET_STRING);
+
+      if (!foundValidSource) {
+        valid = false;
+        const location = {
+          start: {
+            line: (result.line ?? 0) + 1,
```
[the remark this answers is not in the document]
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts]: Added small fix for this.

File: development/webpack/sourcemap-validator.ts [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+  try {
+    rawSourceMap = await readFile(mapPath, 'utf8');
+  } catch {
+    console.error(
+      `SourcemapValidator (webpack) - failed to load sourcemap for "${label}"`,
+    );
+    return false;
+  }
+
+  const consumer = await new SourceMapConsumer(rawSourceMap);
```
[the remark this answers is not in the document]
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts]: Added fix for this.

File: development/webpack/sourcemap-validator.ts [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+ */
+export function isLikelyCommentLine(line: string): boolean {
+  const trimmed = line.trim();
+  return (
+    trimmed.startsWith('*') ||
+    trimmed.startsWith('//') ||
+    trimmed.startsWith('/*') ||
+    trimmed.startsWith('*/') ||
+    trimmed === '' ||
+    /^\s*$/u.test(line)
```
[the remark this answers is not in the document]
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts]: Added fix for this.

File: development/webpack/sourcemap-validator.ts [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+  try {
+    rawSourceMap = await readFile(mapPath, 'utf8');
+  } catch {
+    console.error(
+      `SourcemapValidator (webpack) - failed to load sourcemap for "${label}"`,
+    );
+    return false;
+  }
+
+  const consumer = await new SourceMapConsumer(rawSourceMap);
```
[the remark this answers is not in the document]
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts]: Updated / refactored.

File: development/webpack/sourcemap-validator.ts [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+    chromeDirExists = st.isDirectory();
+  } catch {
+    // ENOENT or other; treat as missing
+  }
+
+  if (!chromeDirExists) {
+    console.error(
+      `SourcemapValidator (webpack) - dist/chrome/ does not exist or is not a directory. Run a webpack build first (e.g. yarn webpack). Exiting with code 1.`,
+    );
+    process.exit(1);
```
[the remark this answers is not in the document]
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts:115]: Adding returns after process.exit probably doesn't make sense, also TypeScript is saying: `TS7027: Unreachable code detected.`.

File: development/webpack/sourcemap-validator.ts [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+ * @returns Array of 0-based indices; empty if substring is not found.
+ */
+export function indicesOf(substring: string, str: string): number[] {
+  const a: number[] = [];
+  let i = -1;
+  while ((i = str.indexOf(substring, i + 1)) >= 0) {
+    a.push(i);
+  }
+  return a;
+}
```
[the remark this answers is not in the document]
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts:381]: Entire browserify setup will eventually be deprecated and deleted, once webpack setup is fully functional and integrated.

File: types/babel__code-frame.d.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
+declare module '@babel/code-frame' {
```
Comment (davidmurdoch) [inline types/babel__code-frame.d.ts]: Will `yarn add --dev @types/babel__code-frame` work instead?

File: development/webpack/sourcemap-validator.ts [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+    try {
+      await main();
+    } catch (error) {
+      console.error(error);
+      process.exit(1);
+    }
+  }
+}
+
+void runEntryPoint();
```
Comment (davidmurdoch) [inline development/webpack/sourcemap-validator.ts]: Renaming this file and your test file to have a `.mts` extension and then doing

```
if (import.meta.main) {
  main();
}
```

Should work, I think? That way we won't need this `env` var, which was a bit confusing at first read.

You'll also need to change the `test:unit:webpack` scripts to include the mts extension (soemthing like `tsx --test development/webpack/test/*.test.{ts,mts}`)

**Review by davidmurdoch — CHANGES_REQUESTED** [review]
just a quick, first glance and partial review

File: types/babel__code-frame.d.ts (hunk 7 above)
Comment (david0xd) [reply types/babel__code-frame.d.ts]: Yes, this works! Updated.

File: development/webpack/sourcemap-validator.ts (hunk 8 above)
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts]: I've tried to do this. But, I couldn't make it work with coverage. It just becomes omitted from the coverage report. I don't see a way how to make this work with the coverage report.

File: development/webpack/sourcemap-validator.ts [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+            },
+          };
+          const codeSample = codeFrameColumns(sourceContent, location, {
+            message: `expected to see ${JSON.stringify(TARGET_STRING)}`,
+            highlightCode: true,
+          });
+          console.error(
+            `Sourcemap seems invalid, ${result.source}\n${codeSample}`,
+          );
+        }
```
[the remark this answers is not in the document]
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts:345]: I think that this is not entirely true. In my opinion this doesn't look like "false validation failure" 🤔 

Added specific error logging for this case, plus the unit test for it. 

P.S. I haven't encountered this case so far while I was testing this.

File: development/webpack/sourcemap-validator.ts [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+          );
+          continue;
+        }
+
+        const sourceLines = sourceContent.split('\n');
+        const sourceLineIndex = (result.line ?? 1) - 1;
+        const sourceLine = sourceLines[sourceLineIndex];
+        const column = result.column ?? 0;
+        const portion = sourceLine ? sourceLine.slice(column) : '';
+        const foundValidSource = portion.includes(TARGET_STRING);
```
[the remark this answers is not in the document]
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts:328]: This would be extreme edge case. Added some handling for this, just in case.

File: development/webpack/sourcemap-validator.ts [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+          source: origSource,
+        } = result;
+        const sourceContent = consumer.sourceContentFor(origSource);
+        if (sourceContent === null) {
+          valid = false;
+          console.error(
+            `SourcemapValidator (webpack) - no source content for "${origSource}", in bundle "${label}"`,
+          );
+          continue;
+        }
```
[the remark this answers is not in the document]
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts:321]: Updated to: `consumer.sourceContentFor(origSource, true)`.

File: development/webpack/sourcemap-validator.ts (hunk 8 above)
Comment (davidmurdoch) [reply development/webpack/sourcemap-validator.ts]: Ah yeah, bummer.

Using env vars here feels like the wrong tool to me. `SOURCEMAP_VALIDATOR_SKIP_MAIN` changes module behavior without appearing in any functions. it’s mutable (from any code in the process), so tests/importers could accidentally affect each other, it couples unrelated code: any thing that imports this must know about it before import so it can set up the env properly. It’s "stringly"-typed (😆) so '0',  '1', 'true', 'false', etc are all truthy here.

I think it'd be better/easier to do something _like_:

use something like
```
// sourcemap-validator.cli.ts
import { runWebpackSourceMapValidator } from './sourcemap-validator';

void runWebpackSourceMapValidator().catch((error) => {
  console.error(error);
  process.exit(1);
});
```

and use that file in the `package.json` script.

What do you think?

Comment (david0xd) [reply development/webpack/sourcemap-validator.ts]: Yes. I was thinking of adding additional file as an entry point, which is probably solid tradeoff in this case. 

So, added it now and removed the env var from the run script.

**Review by DDDDDanica — DISMISSED** [review]

File: development/webpack/test/sourcemap-validator.test.ts [hunk 12]
```diff
@@ -0,0 +1,502 @@
...
+      mock.method(console, 'log', noop);
+      await main();
+      await rm(emptyChromeDir, { recursive: true, force: true });
+      assert.ok(exitMock.mock.calls.length >= 1);
+      assert.strictEqual(exitMock.mock.calls[0].arguments[0], 1);
+    });
+
+    it('exits with 1 when at least one bundle fails validation', async () => {
+      const failDir = await mkdtemp(
+        join(tmpdir(), 'sourcemap-validator-main-fail-'),
```
Comment (davidmurdoch) [inline development/webpack/test/sourcemap-validator.test.ts:570]: In the other tests you always clean up the `tmpdir` and any files it contains after the tests, even if they fail or time out. Can you do the same pattern here to try to limit the chance that we liter the tmp dir with old files and directories if things do go wrong?

File: development/webpack/test/sourcemap-validator.test.ts [hunk 13]
```diff
@@ -0,0 +1,502 @@
...
+
+  describe('validateBundle', () => {
+    let tmpDir: string;
+
+    beforeEach(async () => {
+      tmpDir = await mkdtemp(join(tmpdir(), 'sourcemap-validator-test-'));
+    });
+
+    afterEach(async () => {
+      await rm(tmpDir, { recursive: true, force: true });
```
Comment (davidmurdoch) [inline development/webpack/test/sourcemap-validator.test.ts:176]: *not a request for change*

fun fact: node has a `mkdtempDisposable` function now.

```
let dir;

beforeEach(async () => {
  dir = await mkdtempDisposable(join(tmpdir(), 'mytest-'));
});

afterEach(async () => {
  await dir?.remove();
});
```

But you can also do neat things like:

```
it("the test", () =. {
   await using dir = await  mkdtempDisposable(join(tmpdir(), 'something_here-'));
   
   await mkdir(join(dir.path, 'dist', 'chrome'), { recursive: true });
   // the rest of the test
}); // `dir.path` and all files in it are automatically deleted!
```

However, we can't easily use it without until https://github.com/MetaMask/metamask-extension/pull/40454 is merged.

File: development/webpack/sourcemap-validator.ts [hunk 14]
```diff
@@ -0,0 +1,337 @@
...
+
+    console.log(`  sampling from ${consumer.sources.length} files`);
+    let sampleCount = 0;
+    let valid = true;
+
+    const buildLines = rawBuild.split('\n');
+    for (let lineIndex = 0; lineIndex < buildLines.length; lineIndex++) {
+      const line = buildLines[lineIndex];
+      const matchIndices = indicesOf(TARGET_STRING, line);
+      for (const matchColumn of matchIndices) {
```
Comment (davidmurdoch) [inline development/webpack/sourcemap-validator.ts:256]: because the default for `valid` is `true`, if there are no `buildLines`, or `matchIndexes.length` ends up being `0`, we can return `true`, even though we didn't find the `TARGET_STRING`. We can't just default to `false` and set it to `true` if we DO find a valid string, since we want to ensure we find _all_ of them.

hmmmm, tricky. what do you think?

File: development/webpack/sourcemap-validator.ts [hunk 15]
```diff
@@ -0,0 +1,337 @@
...
+        const position = { line: lineIndex + 1, column: matchColumn };
+        const result = consumer.originalPositionFor(position);
+
+        if (!result.source) {
+          if (isLikelyCommentLine(line)) {
+            continue;
+          }
```
Comment (davidmurdoch) [inline development/webpack/sourcemap-validator.ts]: there is a risk of `false positives` here, but I think its so small, and the risk of it actually happening AND becoming a problem is so small its fine. Just wanted to mention it so you can make the decision to not handle cases like `/* some comment */throw new Error("an error")`. 

I think its fine though.

**Review by davidmurdoch — CHANGES_REQUESTED** [review]

File: development/webpack/sourcemap-validator.ts (hunk 15 above)
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts]: This is legit concern (edge case). I added handling for that and also for multiline block comments.

File: development/webpack/test/sourcemap-validator.test.ts (hunk 12 above)
Comment (david0xd) [reply development/webpack/test/sourcemap-validator.test.ts:570]: Added similar cleanup for other as well.

File: development/webpack/test/sourcemap-validator.test.ts (hunk 13 above)
Comment (david0xd) [reply development/webpack/test/sourcemap-validator.test.ts:176]: Alright, good to know. I assume we're keeping it as-is for now.

File: development/webpack/sourcemap-validator.ts [hunk 16]
```diff
@@ -0,0 +1,379 @@
...
+          });
+          console.error(
+            `Sourcemap seems invalid, ${origSource}\n${codeSample}`,
+          );
+        }
+      }
+    }
+
+    console.log(`  checked ${sampleCount} samples`);
+    return valid;
```
[the remark this answers is not in the document]
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts:355]: Added additional console log to draw more attention to this (just in case). Previously it was only reporting 0 for sample count.

File: development/webpack/sourcemap-validator.ts [hunk 17]
```diff
@@ -0,0 +1,379 @@
...
+    consumer = await new SourceMapConsumer(rawSourceMap);
+    if (!consumer.hasContentsOfAllSources()) {
+      console.warn(
+        'SourcemapValidator (webpack) - missing content of some sources...',
+      );
+    }
+
+    console.log(`  sampling from ${consumer.sources.length} files`);
+    let sampleCount = 0;
+    let valid = true;
```
[the remark this answers is not in the document]
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts:250]: This is similar to this one: https://github.com/MetaMask/metamask-extension/pull/40025#discussion_r2861602869

Console log is added in the process to indicate what's being validated and how many samples are checked and if samples checked equal zero, we log some stuff additionally to rise awareness.

I believe it's enough.

The concept of this script relies on that target string for now, we don't have other approaches for bundles without "throw new error" and some of these bundles can be almost json-only or some similar stuff where there is probably nothing to validate related to the source maps.

File: development/webpack/sourcemap-validator.ts (hunk 14 above)
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts:256]: It's tricky because it's some sort of tradeoff by design. I improved it compared to the old version. I've added console log for these cases so we're aware of it.

Here is an example of how it works in such cases:

```perl
build "js-vendor_trezor_usb-permissions_js-b1c175a1.ebf5bcb23c64625d7eb3.js"
  sampling from 1 files
  checked 0 samples
SourcemapValidator (webpack) - no "new Error" found in bundle "js-vendor_trezor_usb-permissions_js-b1c175a1.ebf5bcb23c64625d7eb3.js"; nothing to validate.
```

I've also tried to return `false` in such cases, but it makes the whole thing always fail the validation since many of these files don't have `TARGET_STRING`. However, I don't find these files *too important* since they don't have a lot of source code in most cases. It looks like there is significant difference in how webpack works with many of these chunk outputs compared to the browserify builds.

Still, the idea seems to work, it's just coming with some tradeoffs that are maybe little bit different compared to the old version and its tradeoffs.

I've been analyzing these files that don't have `TARGET_STRING` before. And my conclusion was that most of the time they're not that much important for source map validation or not at all. Some are almost pure JSON parsing configurations, while some of the files look just like this:
```javascript
(globalThis["webpackChunk"] = globalThis["webpackChunk"] || []).push([[3028],{

/***/ 833028
() {

/* (ignored) */

/***/ }

}]);
//# sourceMappingURL=3028.51ab8728d080144a053c.js.map
```

I believe it's not that much critical for this script (build process feature).

File: development/webpack/sourcemap-validator.ts [hunk 18]
```diff
@@ -0,0 +1,384 @@
...
+      }
+    }
+
+    console.log(`  checked ${sampleCount} samples`);
+    if (sampleCount === 0) {
+      console.error(
+        `SourcemapValidator (webpack) - no "${TARGET_STRING}" found in bundle "${label}"; nothing to validate.`,
+      );
+    }
+    return valid;
```
[the remark this answers is not in the document]
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts:355]: Yes. It looks like it's a tradeoff we must have. 

I've answered on it here: https://github.com/MetaMask/metamask-extension/pull/40025#discussion_r2863818982

And in other threads like this where I mentioned that I added console log which will log warning for such cases, so we're aware of it.

File: development/webpack/test/sourcemap-validator.test.ts [hunk 19]
```diff
@@ -0,0 +1,597 @@
...
+        errors.push(args.map((a) => String(a)).join(' '));
+      });
+      assert.ok(
+        errors.some(
+          (e) =>
+            e.includes('nothing to validate') || e.includes('no "new Error"'),
+        ),
+        'should log that no samples were found',
+      );
+    });
```
[the remark this answers is not in the document]
Comment (david0xd) [reply development/webpack/test/sourcemap-validator.test.ts:282]: Yes. I see that test is failing. I messed this up accidentally while trying to fix lint...

[the remark this answers is not in the document]
Comment (david0xd) [reply development/webpack/test/sourcemap-validator.test.ts:282]: Fixed now.

File: development/webpack/sourcemap-validator.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts:355]: Yes. It looks like it's a tradeoff we must have. 

I've answered on it here: https://github.com/MetaMask/metamask-extension/pull/40025#discussion_r2863818982

And in other threads like this where I mentioned that I added console log which will log warning for such cases, so we're aware of it.

[the remark this answers is not in the document]
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts:355]: Please stop spamming about things we already know and we're already discussing! ☹️

[the remark this answers is not in the document]
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts:355]: Yes. It looks like it's a tradeoff we must have. 

I've answered on it here: https://github.com/MetaMask/metamask-extension/pull/40025#discussion_r2863818982

And in other threads like this where I mentioned that I added console log which will log warning for such cases, so we're aware of it.

[the remark this answers is not in the document]
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts:355]: Yes. It looks like it's a tradeoff we must have. 

I've answered on it here: https://github.com/MetaMask/metamask-extension/pull/40025#discussion_r2863818982

And in other threads like this where I mentioned that I added console log which will log warning for such cases, so we're aware of it.

[the remark this answers is not in the document]
Comment (david0xd) [reply development/webpack/sourcemap-validator.ts:355]: Yes. It looks like it's a tradeoff we must have. 

I've answered on it here: https://github.com/MetaMask/metamask-extension/pull/40025#discussion_r2863818982

And in other threads like this where I mentioned that I added console log which will log warning for such cases, so we're aware of it.

File: development/webpack/sourcemap-validator.ts (hunk 14 above)
Comment (davidmurdoch) [reply development/webpack/sourcemap-validator.ts:256]: sound like a good trade off. thanks for giving it some thought

**Review by davidmurdoch — APPROVED** [review]

**Review by DDDDDanica — APPROVED** [review]
